File: C/EASY_0206_Reverse_Linked_List/solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList(struct ListNode* head) {
    struct ListNode* prev = NULL;
    struct ListNode* next = NULL;

    /* Thoughts:
        Three Pointers
            - One tracking prev node
            - One tracking current node which will be head
            - One tracking next node

            - set current's next node to prev
                - this breaks a link to the next node in the list
                    so save current node's next first
            - move previous to current node
            - move current node to next node
            - repeat until current node is null
            - previous node will be first pointer of the list
    */
    while(head != NULL){
        next = head->next;
        head->next = prev;
        prev = head;
        head = next;
    }
    return prev;
}

File: Python/EASY_0206_Reverse_Linked_List/solution.py
# 206. Reverse Linked List

# Given the head of a singly linked list, reverse the list, and return the reversed list.

# Example 1:

# Input: head = [1,2,3,4,5]
# Output: [5,4,3,2,1]

# Example 2:

# Input: head = [1,2]
# Output: [2,1]

# Example 3:

# Input: head = []
# Output: []

# Constraints:
#     The number of nodes in the list is the range [0, 5000].
#     -5000 <= Node.val <= 5000

# Follow up: A linked list can be reversed either iteratively or recursively. Could you implement both?

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        prev_node = None
        curr_node = head

        while curr_node:
            next_node = curr_node.next
    
            curr_node.next = prev_node

            prev_node = curr_node
            curr_node = next_node   
        
        return prev_node

